[3.x] Add property name style toggle to Inspector

timothyqiu opened the pull request:
This PR adds a button to the Inspector. It switches among three different property name styles:

| Style | Preview |
| --- | --- |
| Raw | `http_proxy` |
| Capitalized | `HTTP Proxy` |
| Localized | `HTTP 代理` | 

The Inspector defaults to the `interface/inspector/default_property_name_style` editor setting at startup. This setting defaults to "Capitalized" and it replaces the old `interface/inspector/capitalize_properties`.

| Three Menu Items | In Action |
| --- | --- |
| ![ksnip_20220323-104401](https://user-images.githubusercontent.com/372476/159613657-404cb192-b90d-40fe-ad75-eb9c556511ac.png) | ![Peek 2022-03-23 10-41](https://user-images.githubusercontent.com/372476/159613682-293d030b-5333-4526-9576-c97d463eb096.gif) |

Localization is currently disabled for "en":

![ksnip_20220323-104429](https://user-images.githubusercontent.com/372476/159614100-45796462-c577-48ce-b3b2-87604bc2c755.png)

The previous PR added a `interface/editor/translate_properties` setting for all properties. This is now replaced by `interface/editor/localize_settings` to focus on settings. It defaults to `true` because settings are usually accessed via GUI instead of code. You can still access the path via right click menu or tooltip.

Editor settings take a few seconds to make effect:

![Peek 2022-03-19 19-03](https://user-images.githubusercontent.com/372476/159119248-687034ec-3fe0-4723-a64c-21c002e577cf.gif)

The `EditorInspector` defaults to Capitalized. Common usages such as the import dock and the export dialog are set to honor `localize_settings` too. The local stack variables inspector in Debugger is kept Raw.

---

I'll make a PR for `master` after this is discussed.

timothyqiu committed:
Add property name style toggle to Inspector

fire-forge commented:
What do you think about moving the options to this menu instead? The search bar and sub-resource dropdown button having different lengths looks a little odd.

![image](https://user-images.githubusercontent.com/67974470/159145136-95ecb906-8663-456f-afb2-f8a478073434.png)

> If we want to disable the "Localized" option for some languages

Localization could probably be disabled for English since localized and capitalized are the same in that case, right?

timothyqiu commented:
Added a blacklist of languages, currently "en" only. It now disables the menu item and adds a tooltip. It also selects Capitalized on startup if you set Localized in Editor Settings for these languages.

![ksnip_20220320-111322](https://user-images.githubusercontent.com/372476/159147609-8a9b1ddd-fb70-449a-af6d-dc20fe5726bf.png)

---

> The search bar and sub-resource dropdown button having different lengths looks a little odd.

It's also unblanced a few versions ago ;) I tried locally to move the switch into the properties menu. I think it looks a bit lack of context, maybe wording should be changed? I'll include this if you think it's OK.

![ksnip_20220320-120413](https://user-images.githubusercontent.com/372476/159147663-4839b95a-e955-443c-a0f5-aab097389cf4.png)

fire-forge commented:
@timothyqiu yes, I think that's better. 

> I think it looks a bit lack of context, maybe wording should be changed?

You could add text to the separator right above it that says something like "Property Name Style".

timothyqiu commented:
Updated. Moved the menu into the object properties menu. Not sure about the separator, but it can be discussed :)

![ksnip_20220320-155045](https://user-images.githubusercontent.com/372476/159153598-9b3ef584-d2d9-44d9-8458-f56c78df6e0c.png)

akien-mga commented on editor/editor_property_name_processor.cpp:
```diff
@@ -55,20 +77,20 @@ String EditorPropertyNameProcessor::_capitalize_name(const String &p_name) const
-	const String capitalized_string = _capitalize_name(p_name);
-	if (EDITOR_GET("interface/editor/translate_properties")) {
-		return capitalized_string;
+		default: {
```
I'd use `STYLE_RAW` (and then put it first to use the same order as the enum) to make it explicit that all cases are intentionally handled. Otherwise it looks like nothing uses `STYLE_RAW`.

akien-mga commented on editor/editor_property_name_processor.cpp:
```diff
@@ -84,6 +106,8 @@ EditorPropertyNameProcessor::EditorPropertyNameProcessor() {
 	capitalize_string_remaps["adb"] = "ADB";
 	capitalize_string_remaps["ao"] = "AO";
 	capitalize_string_remaps["apk"] = "APK";
+	capitalize_string_remaps["armeabi-v7a"] = "armeabi-v7a";
+	capitalize_string_remaps["arm64-v8a"] = "arm64-v8a";
```
Nitpick but `6` comes before `e` in alphanumeric sorting ;)

akien-mga commented on editor/inspector_dock.cpp:
```diff
@@ -446,8 +461,19 @@ void InspectorDock::update(Object *p_object) {
+	p->add_radio_check_item(TTR("Capitalized"), PROPERTY_NAME_STYLE_CAPITALIZED);
+	p->add_radio_check_item(TTR("Localized"), PROPERTY_NAME_STYLE_LOCALIZED);
+
+	if (!EditorPropertyNameProcessor::is_localization_available()) {
+		const int index = p->get_item_index(PROPERTY_NAME_STYLE_LOCALIZED);
+		p->set_item_disabled(index, true);
+		p->set_item_tooltip(index, TTR("Localization not available for current language."));
+	}
```
That doesn't seem to work for me on Linux with the `en` locale, the option is still enabled.

timothyqiu replied:
https://github.com/godotengine/godot/blob/8ecdb0b6e93e165b030e791366db551c8a5d531f/editor/editor_property_name_processor.cpp#L54-L57

This is based on `interface/editor/editor_language`. I tried launching the editor using `LANG=en_US`, the editor setting is "en", and the option is disabled. What's your `interface/editor/editor_language`?

akien-mga replied:
Hm I just tested again and it's working fine now, weird.

akien-mga commented on editor/inspector_dock.cpp:
```diff
@@ -446,8 +461,19 @@ void InspectorDock::update(Object *p_object) {
-	p->add_separator();
 
+	p->add_separator(TTR("Property Name Style"));
+	p->add_radio_check_item(TTR("Raw"), PROPERTY_NAME_STYLE_RAW);
```
I feel "Raw" might not be very explicit. Maybe "Raw Path" or just "Path"/"Property Path"?

timothyqiu replied:
It's only part of a path, i.e. there is no slash in it, more like a snake cased identifier. I think it's confusing to call it Path style.

akien-mga commented:
Thanks!
